docs/guides/testing-functions.md

```diff
@@ -0,0 +1,173 @@
+# Testing functions
+
+# Functions in the `source/` directory tree
+
+To make testing functions easier, any function that appears in the `source/` directory is put into the global scope. This means that you can directly call these functions from your test case. For example:
+
+Source file: `<project root>/source/utilities/myUtil.brs`
+```brightscript
+function myUtil()
+    return "awesome!"
+end function
+```
+
+Test file:
+```brightscript
+m.describe("myUtil.brs", sub()
+    m.it("returns something awesome", sub()
+        ' We can directly call myUtil, because it's in scope
+        output = myUtil()
+        m.assert.equal(output, "awesome!", "output failed to match")
+    end sub)
+end sub)
+```
+
+# Functions in the `components/` directory tree
+
+Component scope is a bit more complicated, so we don't put files in the `components/` directory tree into the global scope. Because of this, any `.brs` files that are in `components/` will need to be tested by pulling the file in using a `<script>` tag in a component.
+
+There are a couple ways to accomplish this:
+
+## Use a source component
+
+The best way to test `.brs` files in the `components/` tree is to write component-level tests for whichever component(s) consume the `.brs` file in your application. This is good practice because your tests will align more closely to how the code is being used in your actual application.
+
+For example, if we have this file: `<project root>/components/foo-workflow/foo.brs`
+```brightscript
+function foo()
+    return "foo"
+end function
+```
+
+We can test the `foo` function based on how a component (which we'll call `FooComponent`) uses it:
+
+### As a callback
+
+If `FooComponent` uses the `foo` function as a callback:
+```xml
+<?xml version="1.0" encoding="UTF-8"?>
+<component name="FooComponent">
+    <script type="text/brightscript" uri="pkg:/components/foo-workflow/foo.brs" />
+    <interface>
+        <field id="myField" type="string" onChange="foo" />
+    </interface>
+</component>
+```
+
+Then we can test it by modifying `myField` in our test case:
+```brightscript
+m.it("calls foo when myField is modified", sub()
+    ' create a spy
+    spy = _brs_.mockFunction("foo")
+
+    ' create the object
+    component = createObject("RoSGNode", "FooComponent")
+
+    ' modify the field to trigger the callback
+    component.myField = "something new"
+
+    ' assert that it was called
+    m.assert.equal(spy.calls.count(), 1, "expected foo to be called once")
+end sub)
+```
```

Comment:
This is a great example!  Do you think it'd be worth including one where the `foo()` function has a side-effect in `m.top`, so that the test can directly examine the end-state?

Reply:
great idea!